lib/vsc/mympirun/pmi/mpi.py

```diff
@@ -0,0 +1,303 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with vsc-mympirun.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+Base PMI MPI class, all actual classes should inherit from this one
+
+The role of the MPI class is very limited, mainly to provide supported PMI flavour/version
+"""
+from __future__ import print_function
+
+import os
+
+from vsc.utils.fancylogger import getLogger
+from vsc.mympirun.common import MpiKlass, eb_root_version, version_in_range
+from vsc.mympirun.pmi.pmi import PMIv2, PMIxv3
+
+# TODO: should be made generic somehow
+#   pmix3 pmi2 compat libraries might not work with all mpi apps
+#   but one can package the slurm pmi2 libs on different location
+#   ideally, one tests for pmix or not, and somehow tests that the pmi2 libs are not from pmix package
+AUTOMATIC = 'AUTOMATIC'
+PMI2LIBS = {
+    'SLURM2': '/usr/lib64/slurmpmi/libpmi2.so',
+    'SYSTEM2': '/usr/lib64/libpmi2.so',  # possibly/likely pmix
+    'SLURM1': '/usr/lib64/slurmpmi/libpmi.so',
+    'SYSTEM1': '/usr/lib64/libpmi.so',  # possibly/likely pmix
+    AUTOMATIC: '/automatic/via/does/not/exist',  # non-exisiting file picks up some defaults and "stuff works"
+}
+
+
+class MPI(MpiKlass):
+
+    PMI = None
+
+    def __init__(self, options, cmdargs, **kwargs):
+        if not hasattr(self, 'log'):
+            self.log = getLogger(self.__class__.__name__)
+
+        self.options = options
+        self.cmdargs = cmdargs
+
+        super(MPI, self).__init__(**kwargs)
+
+        self.ucx = self.has_ucx()
+        self.hcoll = self.has_hcoll()
+
+        # sanity checks
+        if getattr(self, 'sched_id', None) is None:
+            self.log.raiseException("__init__: sched_id is None (should be set by one of the Sched classes)")
+
+        if not self.cmdargs:
+            self.log.raiseException("__init__: no executable or command provided")
+
+    def main(self):
+        """Magic now!"""
+        pmicmd, run_function = self.pmicmd()
+
+        cmd = pmicmd + self.cmdargs
+
+        if self.options.print_launcher:
+            print(' '.join(pmicmd))
+            exitcode = 0
+        elif self.options.dry_run:
+            self.log.info("Dry run, only printing generated mpirun command...")
+            print(' '.join(cmd))
+            exitcode = 0
+        else:
+            exitcode, _ = run_function(cmd)
+
+        if exitcode > 0:
+            self.log.raiseException("main: exitcode %s > 0; cmd %s" % (exitcode, cmd))
+
+    def has_ucx(self):
+        """
+        Determine if there is UCX support
+        """
+        root, version = eb_root_version('ucx')
+        if root:
+            self.log.debug("Found UCX root %s version %s", root, version)
+            return True
+        else:
+            self.log.debug("No UCX root / version found")
+            return False
+
+    def has_hcoll(self):
```

Comment:
More code copying :D (you can leave it, just is nice to see you also do it for a change)

Reply:
fixed